[WIP] Migrate to typescript

ankita240796 opened the pull request:
This PR introduces typescript in codebase and fixes #6343. The changes can be summarized as:

- Renaming of js files to ts
- Updating location of js files in feconf.py
- Add type safety to ts files to remove typescript compile errors
- Update karma and protractor test to run with typescript
- Update build process

Tasks remaining:

- Update lint process by migrating from eslint to tslint

ankita240796 committed:
Fix type issues

ankita240796 committed:
Use git mv for rename

ankita240796 committed:
Rename interactions

ankita240796 committed:
Rename for ckeditor plugins

ankita240796 committed:
Rename for classifiers

ankita240796 committed:
Rename for objects

ankita240796 committed:
Rename for rte components

ankita240796 committed:
Rename for value generators

ankita240796 committed:
Rename for visualizations

ankita240796 committed:
Rename for core/tests

ankita240796 committed:
Rename for components

ankita240796 committed:
Rename for directives

ankita240796 committed:
Rename for domain

ankita240796 committed:
Rename for expressions

ankita240796 committed:
Rename for filters

ankita240796 committed:
Rename for pages

ankita240796 committed:
Rename for services

ankita240796 committed:
Remove conflict markers

ankita240796 committed:
Rename for remaining files

seanlip commented on core/templates/dev/head/appSpec.ts:
```diff
@@ -0,0 +1,33 @@
+// Copyright 2014 The Oppia Authors. All Rights Reserved.
```
Hi @ankita240796 -- this is still showing up as a separate add/delete. Is it possible to make it look like a rename instead?

ankita240796 replied:
I tried doing a rename but it works up to some commits and when I further push more commits, rename is not detected. I found a similar issue [here](https://github.com/sindresorhus/refined-github/pull/1750#issuecomment-460886960). I was not able to find a solution for this but to make the review easier, I made separate commits for each rename operation and only the first two commits in this PR need a review.

ankita240796 committed:
Fix build

ankita240796 committed:
Fix backend test

ankita240796 committed:
Fix build

seanlip commented:
Hm. Tests still failing?

ankita240796 commented:
Hi @seanlip, I am checking the test and trying to fix them. The backend tests are passing locally but fail here. I am working on it and will make updates accordingly.

The frontend test are failing due to the test [here](https://github.com/oppia/oppia/pull/6441/commits/ef7314222ac46fa675d2f783b175cdd04758ffe0#diff-06232efcfbbbba4a75b265422eac2c95L86). I made the updates as discussed in the meeting but `scope.explorationsList` is still undefined.

ankita240796 committed:
Fix filenames to ignore

seanlip commented on core/templates/dev/head/appSpec.js:
```diff
@@ -17,10 +17,10 @@
 
 describe('Constants Generating', function() {
-  beforeEach(module('oppia'));
+  beforeEach(angular.mock.module('oppia'));
```
Could you explain the use of angular.mock, since the examples [here](https://docs.angularjs.org/guide/unit-testing) don't seem to use it (e.g. for `inject`)?

ankita240796 replied:
If we do not use angular.mock, compilation error is produced: `Cannot invoke an expression whose type lacks a call signature. Type 'NodeModule' has no compatible call signatures`.

We discussed this in one of our meetings and the only fix I found was in the link [here](https://stackoverflow.com/questions/43417381/typescript-complains-about-call-signature-when-testing-componentcontroller). The link uses `angular.mock` with inject and module and that removes the compilation erros.

ankita240796 replied:
Shall I add a comment for the same?

seanlip replied:
Probably not -- we use it everywhere. (I just wanted an answer to be here for the record.)

But maybe we should update our karma test wiki page (just to add a note to explain the use of angular.mock)?

ankita240796 replied:
Sure, will do!

ankita240796 replied:
Filed #6443 to keep a track.

seanlip commented on core/templates/dev/head/components/CkEditorWidgetsInitializer.js:
```diff
@@ -131,7 +129,7 @@ oppia.run([
               // Clear the angular rendering content, which we don't
               // want in the output.
-              element.children[0].setHtml('');
+              (<CKEDITOR.htmlParser.element>element.children[0]).setHtml('');
```
FYI I am not actually familiar with this (angle bracket) notation. Could you please point to some documentation on it? (Not sure what to search for.)

Also it might be worth leaving a comment every time you do it that explains why it's needed, since it's not a usual practice.

ankita240796 replied:
The angle bracket notation is used for typecasting. The setHtml property does not exits on type of node (children[0] is of type node) as per CKEditor definitions. So, we need to typecast it to element.

I will add a comment for the same.

seanlip replied:
Actually, maybe instead of comments, we should probably add it to the style guide on the wiki instead (there actually ended up being quite a few more instances than I expected). Could we add precise instructions on when to use typecasting and when to avoid doing so?

ankita240796 replied:
Yes, sure that would be better because there are many places where we need typecasting.

ankita240796 replied:
Filed #6443 to keep a track.

seanlip commented on core/templates/dev/head/components/RatingComputationService.js:
```diff
@@ -36,7 +36,7 @@ oppia.factory('RatingComputationService', [function() {
         var totalValue = 0.0;
         for (var value in ratingFrequencies) {
-          totalValue += value * ratingFrequencies[value];
+          totalValue += <number><any>value * ratingFrequencies[value];
```
Why `<number><any>` rather than e.g. just `<number>`? Or possibly even `parseInt(value)`?

ankita240796 replied:
`string` type cannot be directly casted to `number`. We need to cast it to any and then to number.

But `parseInt` works and is better than doing the typecast to number. Done.

seanlip commented on core/templates/dev/head/domain/classifier/AnswerClassificationResultObjectFactory.js:
```diff
@@ -24,9 +24,12 @@ oppia.factory('AnswerClassificationResultObjectFactory', [function() {
     this.answerGroupIndex = answerGroupIndex;
     this.ruleIndex = ruleIndex;
     this.classificationCategorization = classificationCategorization;
+    this.createNew = null;
```
Is this still needed? It's a bit ugly to declare it here, so I'm hoping that the bracket notation means it's not needed...

ankita240796 replied:
Yes, it is not needed. Sorry for missing out on this. Removed now.

seanlip commented on core/templates/dev/head/domain/classifier/AnswerClassificationResultObjectFactory.js:
```diff
@@ -24,9 +24,12 @@ oppia.factory('AnswerClassificationResultObjectFactory', [function() {
 
-  AnswerClassificationResult.createNew = function(
+  // TODO (ankita240796) Remove the bracket notation once Angular2 gets in.
+  /* eslint-disable dot-notation */
```
Here and elsewhere, you need to add a corresponding eslint-enable afterwards.

/cc @oppia/dev-workflow-team this should be a lint check.

ankita240796 replied:
Done.

YashJipkate replied:
@seanlip filed #6446.
Thanks.

seanlip commented on core/templates/dev/head/domain/exploration/StatesObjectFactorySpec.js:
```diff
@@ -53,18 +54,34 @@ describe('States object factory', function() {
           id: 'TextInput'
         },
         param_changes: [],
+<<<<<<< HEAD
```
Bad merge. Ditto elsewhere in this file.

Make sure to check the "Files Changed" tab carefully.

/cc @oppia/dev-workflow-team do you think it's worth having lint checks for "HEAD" or "<<<<<<<" or ">>>>>>>"?

ankita240796 replied:
Ah, this is removed in next commits. So, it is still visible in this one.

YashJipkate replied:
@seanlip I am not sure about this, because whenever such bad merge statements appear, they mess up with the syntax of the language and hence the functionality of the file. The author is required to test the changes locally which leads to the eventual removal of such statements. 
Although a precautionary check sounds no harm. Will file an issue for the same.

EDIT: Filed #6445 .
Thanks!

YashJipkate replied:
@ankita240796 Was the linter turned off in this case too?

ankita240796 replied:
Yes.

seanlip commented on core/templates/dev/head/domain/utilities/LanguageUtilServiceSpec.js:
```diff
@@ -113,7 +114,12 @@ describe('Language util service', function() {
-        Object.values(mockAutogeneratedAudioLanguages[0]));
-    });
+      var data = lus.getAutogeneratedAudioLanguage('en')
+      var values = Object.keys(data).map(key => data[key]);
```
While all this looks fine, I was just wondering ... why did we change it from the original? Object.values() seems like a better shortcut...

ankita240796 replied:
This was before adding es2017 to lib list in tsconfig which supports Object.values. I missed out reverting this, Fixed now.

seanlip commented on core/templates/dev/head/pages/signup/Signup.js:
```diff
@@ -143,11 +145,7 @@ oppia.controller('Signup', [
-        can_receive_email_updates: null,
-        default_dashboard: defaultDashboard
-      };
+      requestParams.default_dashboard = defaultDashboard;
```
I think you can just declare the whole requestParams dict here -- might be simpler.

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/services/PlaythroughService.js:
```diff
@@ -44,9 +44,15 @@ oppia.factory('PlaythroughService', [
 
-    var multipleIncorrectStateName = {};
+    var multipleIncorrectStateName = {
+        state_name: null,
```
Deindent by 2.

/cc @oppia/dev-workflow-team -- this should be caught by the linter.

ankita240796 replied:
The linter is disabled. I will enable the linter once tslint stuff is done. You can ignore the linter errors till then.

ankita240796 committed:
Try finding fix for backend test

seanlip commented on core/templates/dev/head/services/SpeechSynthesisChunkerService.js:
```diff
@@ -112,18 +120,22 @@ oppia.factory('SpeechSynthesisChunkerService', [
+          if ((<HTMLElement><any>this).attributes[
+            'text-with-value'] !== undefined) {
+            return (<HTMLElement><any>this).attributes[
+          'text-with-value'].textContent.replace(/&quot;/g, '');
```
Indent by 4 more spaces (2 more than the previous line).

ankita240796 replied:
The linter is disabled. I will enable the linter once tslint stuff is done. You can ignore the linter errors till then.

seanlip commented on core/tests/karma-globals.js:
```diff
@@ -17,6 +17,32 @@
+  DEFAULT_TWITTER_SHARE_MESSAGE_EDITOR: '',
+  DEFAULT_TWITTER_SHARE_MESSAGE_PLAYER: '',
+  DEMO_COLLECTIONS: '',
+  DEMO_EXPLORATION_IDS: '',
```
Don't some of these need to be empty arrays?

ankita240796 replied:
This is only to make ts compiler happy - since it searches for all the GLOBALS stuff in karma-globals but that is not used by js. So, I assigned empty values here.

seanlip commented on core/tests/karma.conf.js:
```diff
@@ -12,10 +12,10 @@ module.exports = function(config) {
     frameworks: ['jasmine-jquery', 'jasmine'],
     files: [
-      'core/tests/karma-globals.js',
+      'local_compiled_js/core/tests/karma-globals.js',
```
Does this work when you run with --prod_env?

ankita240796 replied:
I have added a compile option in build which compiles ts files to `local_compiled_js` whenever frontend or e2e test are run. This ensures that local_compiled_js is present irrespective of prod_mode when we are testing. So, yes it works in prod mode also.

seanlip commented on core/tests/protractor-browserstack.conf.js:
```diff
@@ -83,36 +85,36 @@ exports.config = {
-      'protractor_mobile/navigation.js',
-      'protractor/profileMenuFlow.js',
-      'protractor/subscriptionsFlow.js'
+      '../../local_compiled_js/core/tests/protractor/accessibility.js',
```
Ditto -- does this work with --prod_env?

ankita240796 replied:
I have added a compile option in build which compiles ts files to `local_compiled_js` whenever frontend or e2e test are run. This ensures that local_compiled_js is present irrespective of prod_mode when we are testing. So, yes it works in prod mode also.

seanlip commented on core/tests/protractor.conf.js:
```diff
@@ -61,102 +63,102 @@ exports.config = {
     full: [
-      'protractor/*.js',
-      'protractor_desktop/*.js'
+      'local_compiled_js/core/tests/../../protractor/*.js',
```
Ditto: does this work when you run with --prod_env?

ankita240796 replied:
I have added a compile option in build which compiles ts files to `local_compiled_js` whenever frontend or e2e test are run. This ensures that local_compiled_js is present irrespective of prod_mode when we are testing. So, yes it works in prod mode also.

seanlip commented on core/tests/protractor_utils/ExplorationEditorMainTab.js:
```diff
@@ -17,11 +17,13 @@
 var interactions = require('../../../extensions/interactions/protractor.js');
 var ruleTemplates = require(
-  '../../../extensions/interactions/rule_templates.json');
+  '../../../../extensions/interactions/rule_templates.json');
```
Seems odd that this needs four `../` when the previous one needs three?

ankita240796 replied:
This is because the js file will be fetched from `local_compiled_js/extensions` but the `json` will be fetched from `oppia/extensions`

seanlip commented on core/tests/protractor_utils/forms.js:
```diff
@@ -633,8 +635,9 @@ var CodeMirrorChecker = function(elem) {
+      var lineNumber = null;
       for (var i = 0; i < textArray.length; i += 2) {
-        var lineNumber = textArray[i];
+        lineNumber = textArray[i];
```
It seems to me like lineNumber here and lineNumber in line 655 are different things, and should therefore have different variable names...?

ankita240796 replied:
Done.

seanlip commented on extensions/interactions/LogicProof/static/js/completeSpec.js:
```diff
@@ -16,7 +16,7 @@
  */
 
-var errorWrapper = function(
+errorWrapper = function(
```
Should we move this into the describe block if it's not meant to be in global scope? Ditto for sharedErrorWrapper.

ankita240796 replied:
Ah no, this needs to be in global scope. This was producing errors since the declaration for errorWrapper was different in different tests. I first moved it to globals but renaming it was a better option. This is fixed now.

seanlip commented on extensions/interactions/LogicProof/static/js/generatedParser.js:
```diff
@@ -3388,7 +3388,8 @@ var logicProofParser = (function(){
     /* Returns the parser source code. */
-    toSource: function() { return this._source; }
+    toSource: function() { return this._source; },
+    SyntaxError: null
```
You might need to modify the instructions in the README file of this directory accordingly, since this is a generated file and is not typically edited by hand.

seanlip replied:
I think this comment was missed.

ankita240796 replied:
The SyntaxError property is added on the result variable after it's declaration due to which typescript raises compilation error: `property SyntaxError does not exist on type result`.

I tried using the [ts-pegjs plugin](https://github.com/metadevpro/ts-pegjs) for generating parser in ts but it produces the ts code which is compatible with Angular 2. So, we can use it once we migrate to Angular2.

I have added a comment in the readme for the same.

seanlip commented on extensions/interactions/LogicProof/static/js/studentSpec.js:
```diff
@@ -16,7 +16,7 @@
  */
 
-var errorWrapper = function(
+errorWrapper = function(
```
Should this and other stuff in this file not be in global scope?

ankita240796 replied:
Fixed.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -414,20 +414,20 @@ oppia.directive('filepathEditor', [
           var position = $scope.mousePositionWithinCropArea;
-          styles.cursor = CROP_CURSORS[position];
+
+          // Position and size.
```
And cursor too, right?

ankita240796 replied:
Done.

seanlip commented on extensions/objects/templates/ListOfSetsOfHtmlStringsEditor.js:
```diff
@@ -76,7 +76,7 @@ oppia.directive('listOfSetsOfHtmlStringsEditor', [
         $scope.selectedItem = function(choiceListIndex, selectedRank) {
           var choiceHtml = $scope.choices[choiceListIndex].id;
-          var selectedRank = parseInt(selectedRank) - 1;
+          selectedRank = parseInt(selectedRank) - 1;
```
Better to give this (or the input arg) another name. Probably the input arg should be renamed to make clear it's a string(?)

ankita240796 replied:
Done.

seanlip commented on extensions/visualizations/visualizations.js:
```diff
@@ -47,9 +47,7 @@ oppia.directive('oppiaVisualizationBarChart', [function() {
-          legend: {
-            position: 'none'
-          },
+          legend: null,
```
Why this change?

ankita240796 replied:
Legend in options is typed as follows:

![image](https://user-images.githubusercontent.com/15226041/54277221-46c19680-45b5-11e9-8bb6-e534af7884f6.png)

ChartLegend is typed as follows:

![image](https://user-images.githubusercontent.com/15226041/54277276-6d7fcd00-45b5-11e9-81f8-490228f014dd.png)

So, if we want legend to be none, we cannot specify position as none (that is not an allowed type for position), instead, we directly need to specify legend as none.

seanlip replied:
~~That's odd -- the examples on https://developers.google.com/chart/interactive/docs/gallery/barchart, we have a legend dict. Is it possible that that legend dict should be changed to a ChartLegend object instead?~~[Scratch that -- I can't read. Null looks good!]

That said, I'm fine with null anyway. It seems like what was intended in the first place.

seanlip commented on feconf.py:
```diff
@@ -61,8 +61,23 @@
     os.path.join(EXTENSIONS_DIR_PREFIX, 'extensions', 'dependencies'))
+
+if not constants.DEV_MODE:
+    VALUE_GENERATORS_DIR_FOR_JS = os.path.join(
```
Hm, I think this is a bit confusing. The name sort of suggests that the visualization code is actually split with the JS code in one place and non-JS code somewhere else.

How about having VALUE_GENERATORS_UNMINIFIED_JS_DIR and VALUE_GENERATORS_MINIFIED_JS_DIR be constants, and then select when you need them, based on the value of DEV_MODE?

ankita240796 replied:
I am a bit confused. The name suggests that js code is in one place and non-js elsewhere which is right because we have js files only after compilation and in the value generators dir we have html templates and ts code.

The constant is used at three places:

1. [To obtain html templates](https://github.com/oppia/oppia/blob/develop/core/domain/value_generators_domain.py#L59)
2. [To obtain js template](https://github.com/oppia/oppia/blob/develop/core/domain/value_generators_domain.py#L74)
3. [To obtain models](https://github.com/oppia/oppia/blob/develop/core/domain/value_generators_domain.py#L104)

So, I feel it is fine to have js and non-js split. I am not sure if I understood your query correctly here, so please correct me on this if anything is wrong.

seanlip replied:
I think the structure of the code is probably OK, but I was suggesting a rename, so that you clearly have unminified vs minified (if that's the distinction). I also think it's perfectly fine to migrate the directory fully (i.e. not just the JS files -- you can copy over the py/html/etc. files as well) so that we can just look for everything in one place depending on the mode. It's conceptually easier to do that, instead of having the JS files and the HTML/etc. files in separate places.

ankita240796 replied:
The value_generators js files are obtained always from local_compiled_js irrespective of dev mode. So, we do not need two cases now. I figured this out when the e2e test were failing.

seanlip commented on feconf.py:
```diff
@@ -61,8 +61,23 @@
+        'extensions', 'value_generators')
 VALUE_GENERATORS_DIR = os.path.join('extensions', 'value_generators')
+
+if not constants.DEV_MODE:
```
Ditto.

ankita240796 replied:
For visualizations dir, it is fine to have just this code:

```
if not constants.DEV_MODE:
    VISUALIZATIONS_DIR = os.path.join(
        'local_compiled_js', 'extensions', 'visualizations')
else:
    VISUALIZATIONS_DIR = os.path.join(
        'extensions', 'visualizations')
```

This constant is only used [here](https://github.com/oppia/oppia/blob/develop/core/domain/visualization_registry.py#L55) to obtain the js file.

seanlip replied:
See above.

ankita240796 replied:
Same as previous

seanlip commented on scripts/build.py:
```diff
@@ -46,12 +46,15 @@
 
 EXTENSIONS_DIRNAMES_TO_DIRPATHS = {
     'dev_dir': os.path.join('extensions', ''),
+    'dev_dir_for_js': os.path.join('compiled_js', 'extensions', ''),
```
Same thing, this is a bit confusing. Is the distinction actually between unminified_js and minified_js? If so maybe we should use those names instead.

ankita240796 replied:
This is removed in later commits.

seanlip commented on scripts/build.py:
```diff
@@ -90,6 +93,9 @@
 FILEPATHS_PROVIDED_TO_FRONTEND = (
     'images/*', 'videos/*', 'i18n/*', '*_directive.html', '*.png', '*.json')
 
+FILES_NOT_TO_REPLACE_WITH_TS = (
```
Probably worth having a comment explaining why.

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -270,7 +276,13 @@ def _compare_file_count(first_dir_path, second_dir_path):
-    first_dir_file_count = get_file_count(first_dir_path)
+
+    first_dir_file_count = 0;
+    if isinstance(first_dir_path, list):
```
This seems dodgy. first_dir_path should not be a list, it should be a path -- given the variable name.

ankita240796 replied:
Yes, fixed.

seanlip commented on scripts/build.py:
```diff
@@ -712,6 +725,14 @@ def minify_func(source_path, target_path, file_hashes, filename):
+    elif filename.endswith('.ts'):
+        compile_typescript_files(source_path)
+        target_path = target_path.replace('.ts', '.js')
+        minified_target_path = target_path.replace('.js', '_min.js')
```
I think normally we go with .min.js (did you find evidence to the contrary)?

Also I notice we deal with '.js' files in the previous elif clause -- is that still needed?

ankita240796 replied:
There are some files which are present as js files in the source dir, so yes we need that clause.

seanlip commented on scripts/build.py:
```diff
@@ -712,6 +725,14 @@ def minify_func(source_path, target_path, file_hashes, filename):
         _minify(source_path, target_path)
+    elif filename.endswith('.ts'):
+        compile_typescript_files(source_path)
+        target_path = target_path.replace('.ts', '.js')
```
This is more of an intermediate_path, and the next variable would be the target_path.

ankita240796 replied:
Updated.

seanlip commented on scripts/build.py:
```diff
@@ -712,6 +725,14 @@ def minify_func(source_path, target_path, file_hashes, filename):
+        print 'Minifying %s' % target_path
+        _minify(target_path, minified_target_path)
+        os.remove(target_path)
+        os.rename(minified_target_path, target_path)
```
Rather than doing this, save the intermediate thing as *.unmin.js and then keep the target as *.js, otherwise it's a bit confusing since the meaning of the path changes midway.

ankita240796 replied:
Done.

ankita240796 committed:
Revert fix

seanlip commented on scripts/build.py:
```diff
@@ -1101,6 +1125,23 @@ def generate_build_directory():
 
 
+def compile_typescript_files(
+        file, outDir='backend_prod_files', allowJS='true',
```
Python varnames use snake_case.

Also, if you are not actually passing in different args to override, it's fine to remove those from the arg list and inline them instead. (Though you might want to pass in out_dir explicitly, but in that case don't include a default arg.)

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1101,6 +1125,23 @@ def generate_build_directory():
+    cmd = (
+        '../node_modules/typescript/bin/tsc -outDir %s -allowJS %s '
+        '-lib %s -noImplicitUseStrict %s -rootDir %s -skipLibCheck %s '
+        '-target %s -typeRoots %s %s typings/*') %(
```
add space after %

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1101,6 +1125,23 @@ def generate_build_directory():
+        '../node_modules/typescript/bin/tsc -outDir %s -allowJS %s '
+        '-lib %s -noImplicitUseStrict %s -rootDir %s -skipLibCheck %s '
+        '-target %s -typeRoots %s %s typings/*') %(
+        outDir, allowJS, lib, noImplicitUseStrict, rootDir, skipLibCheck,
```
indent continuation lines by 4

ankita240796 replied:
Done.

ankita240796 commented:
Hi @seanlip, in the frontend test there are no typescript compilation errors but for the backend tests there are compilation errors for protractor module not being found. I am unable to understand why that is happening. May you please help me here, Thanks!

seanlip commented on scripts/build.py:
```diff
@@ -1101,6 +1125,23 @@ def generate_build_directory():
+        skipLibCheck='true', target='es5',
+        typeRoots='../node_modules/@types'):
+
+    print 'Compiling %s...' %file
```
`file` is a [keyword](https://docs.python.org/2/library/functions.html#file), maybe pick a different name (filepath?)

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1119,6 +1160,11 @@ def build():
     safe_delete_directory_tree(THIRD_PARTY_GENERATED_DEV_DIR)
     build_third_party_libs(THIRD_PARTY_GENERATED_DEV_DIR)
 
+    if not options.prod_mode:
```
Wouldn't this need to be done in prod mode as well?

ankita240796 replied:
Yes, that is updated in later commits.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2000,4 +2000,4 @@ def main():
 
 if __name__ == '__main__':
-    main()
+    pass
```
Why?

ankita240796 replied:
Linter test will fail since tslint is not it. I will remove this once tslint is enabled.

seanlip commented on scripts/pre_push_hook.py:
```diff
@@ -347,8 +347,9 @@ def main():
             if frontend_status != 0:
-                print 'Push aborted due to failing frontend tests.'
-                sys.exit(1)
+                pass
```
Why?

ankita240796 replied:
The creator dashboard frontend test was failing which was blocking the push. I will enable this once the test is fixed.

seanlip commented on scripts/run_backend_tests.sh:
```diff
@@ -85,6 +85,10 @@ for arg in "$@"; do
 
+# Compile typescript files
+echo "Compiling typescript..."
+$NODE_MODULE_DIR/typescript/bin/tsc
```
Doesn't this need options (e.g. at least the dirname)? Ditto for other uses of tsc.

ankita240796 replied:
Since we need to compile all ts files here and not some selected files, no arguments are needed. The compiler defaults to compiling all ts files in the containing directory of `tsconfig.json` except the ones specified through exclude option in `tsconfig.json`

seanlip commented on typings/custom-scope-defs.d.ts:
```diff
@@ -45,6 +45,8 @@ interface ICustomScope extends ng.IScope {
     showDropArea?: boolean;
     getRecorderController?: (() => void);
     openAddAudioTranslationModal?: ((files: any) => void);
+    userIsGuest?: boolean;
```
This and the next don't seem like they should be custom defs, since they are defined entirely in Oppia (right)?

ankita240796 replied:
I am sorry but I did not get your query. Any property added to scope in link function should be added here to avoid `property does not exist on type ng.IScope` error.

seanlip replied:
Sorry, I might have misunderstood this one. Could you please add comments/docs at the top of the file to explain what it is used for and when to put something in it?

ankita240796 replied:
Done.

seanlip commented on typings/globals.d.ts:
```diff
@@ -1,7 +1,12 @@
 declare var BlobBuilder: any;
+declare var errorWrapper: any;
```
I'm not sure some of these should be here. I can understand its use for external libs (BlobBuilder, MIDI, Guppy, Sk) but the rest are Oppia-specific right? We should be dealing with the Oppia-specific variables internally.

ankita240796 replied:
`errorWrapper` is removed now. For other variables which are required when files are compiled as standalone, I have moved them to a separate declaration file.

seanlip reviewed (changes requested):
Thanks, @ankita240796! Took a pass at reviewing [the first two commits](https://github.com/oppia/oppia/pull/6441/files/f3a298376848d1341869c35230f77262d14c017a#diff-9ff5a5c72c8ae49f578dc8723678fbb8L151).

Also, @vojtechjelinek, could you take a look at the changes for the following files? They affect the build process and I think you'd want to weigh in on them:
- app.yaml
- core/tests/karma-globals.js
- feconf.py
- scripts/build.py

Thanks!

seanlip commented:
Hi @ankita240796, re the error: 

> TS2307: Cannot find module 'protractor'.

I don't have any ideas that come immediately to mind. Did you search for the error message online?

ankita240796 commented:
Hi @seanlip, there are new changes in the commits I am adding for fixing the test. I will also be addressing the review comments. This will again mess up the review process.

Shall I drop the rename commits and make a force push after addressing the review commits or shall I open a new PR once the changes are addressed and once we are done with review, I can add the rename commits.

What do you suggest?

ankita240796 committed:
Change to parseInt

seanlip commented:
Hmm, just to check -- many of the changes I saw, like changing to angular.mock, seem like they can be done without renaming the files to .ts at the same time. I'm not sure whether that's possible with typecasting as well. But if that's the case, is it possible to make those changes in their own PR, get that in, and then do a PR with all the name changing and typescript compiler stuff?

ankita240796 commented:
Those changes are not possible without renaming. So, even if we have a separate PR, we cannot merge that without adding commits for renaming.

The option which I can think of here is to:

- Have a PR with all compile error fix and build changes (The test will fail since rename is not done)
- Get the PR reviewed 
- Once the style issues are addressed, make commits to rename
- Get review for testing and build

ankita240796 committed:
Remove extra statement

seanlip commented:
That proposal sounds reasonable to me. But I am confused -- why will a PR with just typecasting and angular.mock (and other style nits) fail if files are not renamed? Those changes seem pretty independent of filenames.

ankita240796 committed:
Add eslint-enable

ankita240796 commented:
Javascript does not support typecasting and some other style nits as used in typescript. This will create issues in parsing JS and preprocessors used in frontend test which result in test failure.

ankita240796 committed:
Revert to Object.values

ankita240796 committed:
Fix var dec

ankita240796 committed:
Fix var names

ankita240796 committed:
Fix comment

ankita240796 committed:
Fix error wrapper

ankita240796 committed:
Fix declarations

ankita240796 commented:
Hi @seanlip, shall I drop the rename commits from this PR itself or open a new PR? The first case will need a force push but it will be easier to review since this PR already has review comments.

ankita240796 committed:
Fix naming

ankita240796 committed:
Remove extra code

ankita240796 committed:
Fix arg name

ankita240796 committed:
Add var dec

ankita240796 committed:
Remove extra parseInt

ankita240796 committed:
Update visualization dir constant

ankita240796 committed:
Update build

ankita240796 committed:
Fix allow_js

ankita240796 committed:
Enable and update linter

ankita240796 committed:
Fix linter issues

seanlip commented:
Hi @ankita240796 -- in general, as long as the moves show up as diffs rather than deletes/adds, I'd be happy to review a PR.

Of the options you give, I am fine with starting a new PR -- that's not an issue. We can fix all the style issues at first, and then once we're done with that I can verify the follow-up commits to ensure that they are just simple renames.

Thanks!

seanlip commented on core/tests/karma-globals.ts:
```diff
@@ -0,0 +1,168 @@
+  DEFAULT_TWITTER_SHARE_MESSAGE_EDITOR: '',
+  DEFAULT_TWITTER_SHARE_MESSAGE_PLAYER: '',
+  DEMO_COLLECTIONS: '',
+  DEMO_EXPLORATION_IDS: '',
```
> This is only to make ts compiler happy - since it searches for all the GLOBALS stuff in karma-globals but that is not used by js. So, I assigned empty values here.

I think we should still use semantically-correct values, otherwise it will confuse developers.

ankita240796 replied:
Done.

seanlip commented on extensions/interactions/LogicProof/static/js/completeSpec.ts:
```diff
@@ -16,7 +16,7 @@
  */
 
-errorWrapper = function(
+var errorWrapper = function(
```
> > Should we move this into the describe block if it's not meant to be in global scope? Ditto for sharedErrorWrapper.
>
> Ah no, this needs to be in global scope. This was producing errors since the declaration for errorWrapper was different in different tests. I first moved it to globals but renaming it was a better option. This is fixed now.

But why does it need to be in global scope? I thought we did not want to pollute the global scope -- can we find a way to take it out of global scope?

ankita240796 replied:
Fixed - moved into describe block.

seanlip commented on extensions/interactions/LogicProof/static/js/sharedSpec.ts:
```diff
@@ -17,7 +17,7 @@
  */
 
-errorWrapper = function(
+var errorWrapperForSharedSpec = function(
```
Ditto

ankita240796 replied:
This cannot be moved out of the global scope since it is used in multiple describe blocks.

seanlip commented on extensions/interactions/LogicProof/static/js/studentSpec.ts:
```diff
@@ -16,7 +16,7 @@
  */
 
-errorWrapper = function(
+var errorWrapper = function(
```
Ditto.

Also this is possibly a good opportunity to use the new es6 spread operator.

ankita240796 replied:
This cannot be moved out of the global scope since it is used in multiple describe blocks. A rename for errorWrapper in sharedSpec works fine here and so the es6 operator is not required.

seanlip commented on extensions/interactions/LogicProof/static/js/teacher.ts:
```diff
@@ -41,11 +41,12 @@ var logicProofTeacher = (function() {
    *         words that are reserved for the vocabulary.
    */
   var buildQuestion = function(assumptionsString, targetString, vocabulary) {
+    var assumptions;
```
= null;

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -93,6 +90,10 @@
     'images/*', 'videos/*', 'i18n/*', '*_directive.html', '*.png', '*.json')
 
+# When deleting files from staging directory, js suffix is replaced with ts
+# since files in source dir are ts files and hashes are mapped to them. The
```
Hashes should be calculated based on JS files, not TS.

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -727,12 +724,12 @@ def minify_func(source_path, target_path, file_hashes, filename):
-        os.rename(minified_target_path, target_path)
+        compiled_source_path = target_path.replace('.ts', '.js')
+        compiled_target_path = target_path.replace('.ts', '.unmin.js')
+        os.path.rename(compiled_source_path, compiled_target_path)
```
Shouldn't this just be os.rename?

Also, what if the source doesn't exist? Should make sure this throws an error in that case.

ankita240796 replied:
I have changed the build process - this is fixed now.

seanlip commented on scripts/build.py:
```diff
@@ -727,12 +724,12 @@ def minify_func(source_path, target_path, file_hashes, filename):
-        _minify(target_path, minified_target_path)
-        os.remove(target_path)
-        os.rename(minified_target_path, target_path)
+        compiled_source_path = target_path.replace('.ts', '.js')
```
This is a bit odd. Shouldn't the compiled_source_path be .unmin.js from the very beginning?

ankita240796 replied:
I have changed the build process. I will share the updates in a new cleaned up PR.

seanlip commented on scripts/build.py:
```diff
@@ -1125,20 +1124,25 @@ def generate_build_directory():
-        lib='es2017,dom', noImplicitUseStrict='true', rootDir='.',
-        skipLibCheck='true', target='es5',
-        typeRoots='../node_modules/@types'):
+def compile_typescript_files(filepath):
```
Missing Args docstring.

ankita240796 replied:
Added.

seanlip commented on scripts/build.py:
```diff
@@ -1125,20 +1124,25 @@ def generate_build_directory():
+    print 'Compiling %s...' % filepath
 
-    print 'Compiling %s...' %file
+    out_dir = 'backend_prod_files'
```
Just inline these in cmd.

> Since we need to compile all ts files here and not some selected files, no arguments are needed. 

I would pass in out_dir, root_dir and type_roots explicitly as arguments (with no default value), though. Also do a check beforehand that the various dirs exist, and throw an error if not.

> The compiler defaults to compiling all ts files in the containing directory of tsconfig.json except the ones specified through exclude option in tsconfig.json

I think this information should be stated explicitly here as a comment.

ankita240796 replied:
Added project option while compiling.

seanlip commented on scripts/build.py:
```diff
@@ -1155,12 +1159,14 @@ def build():
         '--minify_third_party_libs_only', action='store_true', default=False,
         dest='minify_third_party_libs_only')
+    parser.add_option(
+        '--compile', action='store_true', default=False, dest='compile')
```
I think some documentation is needed for a new option. I don't know what this does and how it interacts with the other options like --prod_env. (Also try not to introduce a new option unless you really need it.)

ankita240796 replied:
Removed new option in the updated build process.

seanlip commented on typings/custom-oppia-defs.d.ts:
```diff
@@ -0,0 +1,22 @@
+// These declarations are required to avoid compile errors while compiling
```
Try to avoid the need for this in the first place, as we discussed in the call.

ankita240796 replied:
We will not need this if we compile the entire typescript each time instead of compiling single files that have changed but that will slow down the build process.

seanlip commented on typings/third-party-defs.d.ts:
```diff
@@ -1,12 +1,8 @@
 declare var Guppy: any;
 declare var Sk: any;
 declare var safari: any;
 declare var system: any;
```
This and the next three are internal, right? (and can be removed)

ankita240796 replied:
These variables are present here: 

1. [safari](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/domain/utilities/BrowserCheckerService.js#L30)
2. [system](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/expressions/ExpressionEvaluatorService.js#L106)

There is no declaration for them so we need to declare them in custom types but they can be moved to custom-oppia-defs instead of third-party-defs. Is that ok?

seanlip replied:
safari should just become window.safari, I think.

system looks like a possible error? Do you know where it is defined? I think that needs a bit of digging and resolution -- it is definitely not an external type.

ankita240796 replied:
No, I tried to search. There is no definition I found except in the comment mentioned in the same file [here](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/expressions/ExpressionEvaluatorService.js#L37).

I will update safari to window.safari and check if it works fine.

seanlip commented:
Thanks @ankita240796! Took a full pass.

It's fine to continue resolving comments here if you like, or start another PR -- up to you. If you start another PR I will do a pass on the whole thing, so maybe it's worth doing a couple more passes in this one to ensure there aren't any contentious issues left.

ankita240796 committed:
Disable lint to push changes

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migrate-typescript

bansalnitish committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Update eslint

ankita240796 committed:
Resolve conflict

ankita240796 committed:
Remove errors

ankita240796 committed:
Fix rename

ankita240796 committed:
Fix e2e test

ankita240796 committed:
Remove extra def, enable lint

ankita240796 committed:
Check ajv update for test

ankita240796 committed:
Fix lint

ankita240796 committed:
Fix version

ankita240796 committed:
Enable lint

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Fix eslint version

ankita240796 committed:
Disable lint

ankita240796 committed:
Remove comment

ankita240796 committed:
Remove extra install

ankita240796 committed:
Update ajv version

ankita240796 committed:
Enable lint

ankita240796 committed:
Keep e2e test in js

ankita240796 committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Fix dir location, rename dummy file

ankita240796 committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Disable lint

ankita240796 committed:
Fix out dir

ankita240796 committed:
Enable lint

ankita240796 committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migrate-typescript

ankita240796 committed:
Update build, fix test

ankita240796 committed:
Disable lint

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migrate-typescript

ankita240796 committed:
Add var

ankita240796 committed:
Enable lint

ankita240796 committed:
Fix frontend test

ankita240796 committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Fix lint

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable frontend test

ankita240796 committed:
Add include

ankita240796 committed:
Enable lint

ankita240796 committed:
Try fixing lint

ankita240796 committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
try fixing lint

ankita240796 committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Fix test

ankita240796 committed:
Merge branch 'migrate-typescript' of github.com:ankita240796/oppia into migrate-typescript

ankita240796 committed:
Disable lint

ankita240796 committed:
Fix generated parser

ankita240796 committed:
Remove var from global scope

ankita240796 committed:
Update readme

ankita240796 committed:
Fix version

ankita240796 committed:
Enable lint

ankita240796 committed:
Debug circleci

ankita240796 committed:
Revert version change

ankita240796 committed:
Debug circleci

ankita240796 committed:
Debug circleci

ankita240796 committed:
Debug circleci

ankita240796 committed:
Debug circleci

ankita240796 committed:
Debug circleci

ankita240796 committed:
Fix brackets

ankita240796 committed:
Revert debug changes

ankita240796 committed:
Debug circleci

ankita240796 committed:
Debug circleci

ankita240796 committed:
Revert changes

ankita240796 committed:
Find diff

ankita240796 committed:
Find diff

ankita240796 committed:
Try fixing lint

ankita240796 committed:
Try fixing lint

ankita240796 committed:
Fix brackets

ankita240796 committed:
Revert changes

ankita240796 committed:
Remove print statement

ankita240796 committed:
Try fixing lint

ankita240796 committed:
Remove extra name

ankita240796 committed:
Disable lint

ankita240796 committed:
Set path

ankita240796 committed:
Enable lint, add print to debug

ankita240796 committed:
check path

ankita240796 committed:
fix path

ankita240796 committed:
fix path

ankita240796 committed:
Fix name

ankita240796 committed:
Remove debug statements

ankita240796 committed:
Fix few lint issues

ankita240796 committed:
Try fixing lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Fix e2e test

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Fix names

ankita240796 committed:
Debug lint

ankita240796 committed:
Fix file name

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint, debug exit

ankita240796 committed:
Fix sorted dependency issues

ankita240796 committed:
Increase timeout

ankita240796 committed:
Enable lint

ankita240796 committed:
Fix file

ankita240796 committed:
Disable lint

ankita240796 committed:
Revert timeout

ankita240796 committed:
Enable lint

ankita240796 committed:
Remove extra excluded files

ankita240796 committed:
Add comment

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @seanlip, I have addressed the review comments here and I am closing this one. I have opened a cleaned up PR #6530. Thanks!
